### PR TITLE
Working npm_install rule with example and test.

### PR DESCRIPTION
**Pull request by mattmoor**
Fixes: https://github.com/bazelbuild/rules_nodejs/issues/63

### CODE REVIEW

**Comment by mattmoor** [comment]
I think I need to prefix commands with `node`...

File: internal/packages.bzl [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# limitations under the License.
+
+"""Rules to install NodeJS dependencies during WORKSPACE evaluation."""
+
+def _npm_install_impl(repository_ctx):
+  """Core implementation of npm_install."""
+
+  repository_ctx.file("BUILD", """
+package(default_visibility = ["//visibility:public"])
+filegroup(name = "node_modules", srcs = glob(["node_modules/**/*"]))
```
Comment (alexeagle) [inline internal/packages.bzl]: there's a problem: many npm packages ship with files that are not valid labels. sometimes the user currently needs to give specific glob patterns under node_modules which are safe, eg. https://github.com/angular/angular/blob/master/BUILD.bazel#L7

If that's not possible with bazel-managed-deps, could you at least document that in the tradeoffs section in the readme?

File: internal/packages.bzl [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  result = repository_ctx.execute(
+    [repository_ctx.path(repository_ctx.attr._npm),
+     "install", repository_ctx.path("")])
+
+  if result.return_code:
+    fail("npm_install failed: %s (%s)" % (result.stdout, result.stderr))
+
+npm_install = repository_rule(
+    attrs = {
+        "packages": attr.label(
```
Comment (alexeagle) [inline internal/packages.bzl]: a plural name looks like it should be a `label_list`
in other places I call this attribute "package_json" so it's more clear what value it expects

File: examples/packages/WORKSPACE [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    name = "npm_json",
+    packages = "//:package.json",
+)
+
+npm_install(
+    name = "npm_lock",
+    packages = "//:package-lock.json",
+)
+
+# load("@build_bazel_rules_nodejs//:defs.bzl", "yarn_install")
```
Comment (alexeagle) [inline examples/packages/WORKSPACE]: remove commented block, let's file an issue for yarn support in bazel-managed-deps and see if anyone actually wants it

File: examples/packages/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,31 @@
+load("@build_bazel_rules_nodejs//:defs.bzl", "nodejs_binary")
+
+VARIANTS = ["npm_json", "npm_lock"]
```
Comment (alexeagle) [inline examples/packages/BUILD.bazel]: this doesn't look much like an example for users to follow. I suspect there should be some e2e testing folder under `/internal` where this kind of exhaustive checking lives.

The better example would just be to convert `examples/program` to use bazel-managed-deps

File: internal/packages.bzl [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
```
Comment (alexeagle) [inline internal/packages.bzl]: my style convention has been to name the file after the rule it exports, so this should be
npm_install.bzl

**Comment by alexeagle** [comment]
Hmm, here's a complication:

Many bazel rules in the nodejs ecosystem need some runtime dependency on a nodejs library. For example, rules_typescript exposes a bazel worker so it needs protobuf at runtime. In addition, this bazel-to-npm dependency edge needs to carry version restrictions, for example rules_typescript works with typescript@'>=2.4 <2.6'

The way I do this is by installing npm package `@bazel/typescript`, then tell the user to add to `WORKSPACE`
```
local_repository(name = "build_bazel_rules_typescript", path="node_modules/@bazel/typescript")
```

Now if they use bazel-managed dependencies, how will they point bazel to the repository rule that creates the `@build_bazel_rules_typescript` workspace? Same issue for most rules.

**Comment by alexeagle** [comment]
Merged most of this as 2b07bb6853ac2def1a493f251674f3e54d31e475 with a follow-up commit I added for the comments above.

There was still a test failing, which I left a TODO for, if you have time to re-base this on master and turn on the `@packages_example//:test_escape_npm_lock` that would be nice

**Comment by alexeagle** [comment]
Was merged in 2b07bb6853ac2def1a493f251674f3e54d31e475

**Pull request closed without merging**